src/clp_ffi_py/ir/PyLogEvent.hpp

```diff
@@ -0,0 +1,144 @@
+     * given by default. It should be given when deserializing the object from
+     * a saved state.
+     * @return true on success.
+     * @return false on failure with the relevant Python exception and error
+     * set.
+     */
+    [[nodiscard]] auto init(
+            std::string_view log_message,
+            ffi::epoch_time_ms_t timestamp,
+            size_t index,
+            PyMetadata* metadata,
+            std::optional<std::string_view> formatted_timestamp = std::nullopt
+    ) -> bool;
+
+    /**
+     * Validates whether the PyLogEvent has a PyMetadata object associated.
+     * @return Whether `Py_metadata` points to nullptr.
+     */
+    [[nodiscard]] auto has_metadata() -> bool { return nullptr != m_py_metadata; }
+
+    /**
+     * Resets pointers to nullptr.
+     */
+    auto reset() -> void {
+        m_log_event = nullptr;
+        m_py_metadata = nullptr;
+    }
+
+    /**
+     * Releases the memory allocated for underlying metadata field and the
+     * reference hold for the Python object(s).
+     */
+    auto clean() -> void {
+        Py_XDECREF(m_py_metadata);
+        delete m_log_event;
+    }
+
+    /**
+     * Binds the given PyMetadata and holds a reference. If `Py_metadata` has
+     * been set already, decrement the reference to discard the old value.
+     * @param metadata
+     */
+    auto set_metadata(PyMetadata* metadata) -> void {
+        Py_XDECREF(m_py_metadata);
+        m_py_metadata = metadata;
+        if (nullptr != metadata) {
+            Py_INCREF(m_py_metadata);
+        }
+    }
+
+    /**
+     * Gets the formatted log message represented by the underlying log event.
+     * If a specific timezone is provided, this timezone is used to format the
+     * timestamp. If the timezone is not provided (Py_None), the default
+     * formatted timestamp from the log event is used. In the case where
+     * the log event does not have a cached formatted timestamp, it obtains one
+     * using the default timezone from the metadata (if metadata is present),
+     * or defaults to UTC.
+     * @param timezone Python tzinfo object that specifies a timezone.
+     * @return Python string of the formatted log message.
+     * @return nullptr on failure with the relevant Python exception and error
+     * set.
+     */
+    [[nodiscard]] auto get_formatted_message(PyObject* timezone = Py_None) -> PyObject*;
+
+    [[nodiscard]] auto get_log_event() -> LogEvent* { return m_log_event; }
+
+    [[nodiscard]] auto get_py_metadata() -> PyMetadata* { return m_py_metadata; }
+
+private:
+    PyObject_HEAD;
+    LogEvent* m_log_event;
+    PyMetadata* m_py_metadata;
+};
+
+/**
+ * Gets the PyTypeObject that represents PyLogEvent's Python type. This type is
+ * dynamically created and initialized during the execution of
+ * `PyLogEvent_module_level_init`.
+ * @return Python type object associated with PyLogEvent.
+ */
+auto PyLogEvent_get_PyType() -> PyTypeObject*;
```

Comment:
ideally, it could. But conceptually the actual type object returned by this function is not `PyLogEvent` object. It's a type object dynamically created during the module initialization, and not directly correlated to the `PyLogEvent` object. I would prefer to keep them separate.